src/XMonad/Layout.hs

```diff
@@ -247,22 +230,19 @@ instance (LayoutClass l a, LayoutClass r a) => LayoutClass (Choose l r) a where
     handleMessage c@(Choose d l r) m | Just ReleaseResources <- fromMessage m =
         join $ liftM2 (choose c d) (handle l ReleaseResources) (handle r ReleaseResources)
 
-    handleMessage c@(Choose CL _ r) m
-        | Just (JumpToLayout desc) <- fromMessage m
-        , desc == description r
-        = Just <$> swap c
-
-    handleMessage c@(Choose CR l _) m
-        | Just (JumpToLayout desc) <- fromMessage m
-        , desc == description l
-        = Just <$> swap c
-
-    handleMessage c m
-        | Just JumpToLayout{} <- fromMessage m
-        = do m' <- passOnM m c
-             if isNothing m'
-                 then traverse (swap . switch) =<< passOnM m (switch c)
-                 else return m'
+    handleMessage c@(Choose d l r) m | Just (JumpToLayout desc) <- fromMessage m =
+        case d of
+            CL -> if   desc == description r
+                  then choose c CR Nothing Nothing
+                  else handleMessage l m >>= \case
+                      Just ml -> choose c CL (Just ml) Nothing
+                      Nothing -> choose (Choose CR l r) CR Nothing =<< handleMessage r m
```

Comment:
Why do you manually construct `(Choose CR l r)` instead of using `c` here? I think that will suppress the `Hide` events.

Reply:
`c` is of the form `Choose CL l r`, if I wanted to use it I'd need another function `switch`-like function.  Putting `c` there would hide the current layout in every case, even if we're on, say, `Full` and executed `sendMessage (JumpToLayout Full)`

Reply:
If we're on `Full` and we `JumpToLayout Full`, it's handled by the `desc == description r` case above, isn't it?

`choose (Choose CR l r) CR` and `choose (Choose CL l r) CL` will never `hide` anything, and that doesn't sound right.

Reply:
This all makes perfect sense, yet it still works (even with decorated layouts)...

We can, however, at least cover the "we're on the same layout" part by checking for descriptions in either case.  This, like magic, also enables us to get rid of the non-hiding calls to `choose` ( 0084a1af5c08fa3015af5aa2fd45d0c165d86a67 )

Reply:
Can you explain what the problem was with the "we're on the same layout" part? That may have been the bit I missed…

Reply:
> Can you explain what the problem was with the "we're on the same layout" part? That may have been the bit I missed…

You said that 

  > If we're on `Full` and we `JumpToLayout Full`, it's handled by the `desc == description r` case above, isn't it?

but it might well be that our layouts looks like `Choose CL Full (*other-layouts*)` and so we'd have to check `desc == description l` in the `CL` case as well (which we and also the ol LayoutCombinators didn't do).  As it stood, this would have then gone through to the other cases, which then managed to switch the layout (I imagine `handleMessage l m` returned `Nothing` and then we flipped `CL` to `CR` in `choose`)